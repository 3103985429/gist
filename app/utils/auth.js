const jwt = require('jsonwebtoken');
const winston = require('winston');

const Users = require('../models/users');
const Tokens = require('../models/tokens');
const { extractHostname } = require('../../shared/url');
const { UserError } = require('../libs/system/errors');
const { handleError } = require('../utils/requests');

let ldapConfig;

if (process.env.AUTH_PROTOCOL === 'LDAP') {
    try {
        const { ldapConfig: customLdapConfig } = require(`../../user/${process.env.SUBDOMAIN}/src/auth`);
        ldapConfig = Object.assign({}, getDefaultLdapConfig(), customLdapConfig);
    } catch (e) {
        ldapConfig = getDefaultLdapConfig();
    }
}


function getDefaultLdapConfig() {
    return {
        getProfileDetails(profile) {
            let { firstName = '', lastName = '' } = profile;
            const email = profile.email || '';

            if (!firstName && profile.name) {
                [firstName, lastName] = profile.name.split(' ');
            }

            return { firstName, lastName, email };
        },
        getStrategyConfig() {
            return {};
        },
        onAfterLogin(req, res, profile, callback) {
            callback();
        },
    };
}

const appendTokenToRequest = (req, res, next) => {
    try {
        const { token } = req.query;
        req.token = token;

        next();
    } catch (error) {
        winston.error(error);

        return res.status(404).render('error', {});
    }
};

const loginWithToken = async (req, res, next) => {
    try {
        const token = await Tokens.findOne({ token: req.token || req.params.token }).populate('apiKey');

        if (!token) {
            req.flash('error', { message: 'Token is invalid.' });

            return res.redirect('/auth/login');
        }

        const referer = req.header('Referer');

        if (!referer) {
            return handleError(new UserError('Missing referer'), res);
        }

        const isAuthorisedDomain = token.apiKey.requestDomains.some(domain => extractHostname(referer) === domain);

        if (!isAuthorisedDomain) {
            return handleError(new UserError('This domain is not authorised.'), res);
        }

        const decoded = jwt.verify(token.token, process.env.SESSION_SECRET);
        const user = await Users.findById(decoded.userId);

        if (!user.defaultLoginTeam || !user._team.length || token.apiKey.team.toString() !== user.defaultLoginTeam.toString()) {
            req.flash('error', { message: 'User is not associated with a team.' });
            return res.redirect('/auth/login');
        }

        req.logIn(user, async (err) => {
            if (err) {
                return next(err);
            }

            token.usedAt = new Date();
            await token.save();

            req.user = user;

            if (req.token) {
                return next();
            } else {
                return res.redirect('/dashboard');
            }

        });
    } catch (e) {
        req.flash('error', { message: 'Token is invalid.' });
        return res.redirect('/auth/login');
    }
};

module.exports = {
    ldapConfig,
    appendTokenToRequest,
    loginWithToken,
};
