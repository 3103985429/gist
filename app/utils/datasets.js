const { findDataset } = require('../controllers/api/dataset/get-helpers');
const { some, includes } = require('lodash');
const DATA_SOURCE_TYPES = require('../../config/data-source-types');

module.exports.getDataset = getDataset;
module.exports.hasValidDatasetSource = hasValidDatasetSource;

/**
 * @param {String|ObjectId} id
 * @returns {Promise<Object>}
 */
async function getDataset(id) {
    const { description } = await findDataset({ id });

    return description;
}

/**
 * Determins if the datasourceDescription passed has a valid dataset source
 * determined from the list in config/data-source-types.js
 * @param {Object} datasourceDescription
 * @returns {Boolean}
 */
function hasValidDatasetSource(datasourceDescription) {
    return some(datasourceDescription, (value, key) => {
        return includes(DATA_SOURCE_TYPES, key);
    });
}
