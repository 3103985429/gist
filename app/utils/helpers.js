const { isEqual, trim, toString } = require('lodash');

module.exports.equals = a => b => isEqual(a, b);

module.exports.isEmbed = (query) => query.embed === 'true' || query.embed === true;

module.exports.formatColumnName = (value) => trim(toString(value).replace(/\./g, '_'));

module.exports.formatRawValues = (value) => trim(toString(value).replace(/’/g, '\'').replace(/”/g, '"'));

module.exports.addSlugIndex = (slugs, item, tilteProp = 'title') => {
    if (!slugs.length) {
        return;
    }

    let i = 2;
    const slugList = slugs.map((item) => item.slug);
    const max = i + slugList.length + 1;

    for (i; i < max; i++) {
        const newSlug = `${item.slug}-${i}`;

        if (!slugList.includes(newSlug)) {
            item.slug = newSlug;
            item[tilteProp] = `${item[tilteProp]} ${i}`;

            break;
        }
    }
};
