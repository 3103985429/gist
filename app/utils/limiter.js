const RateLimit = require('express-rate-limit');
const MongoStore = require('rate-limit-mongo');
const { limiterStore } = require('./mongoStore');

const dbURI = process.env.MONGODB_URI;

const windowTime = 15 * 60 * 1000; //15 minutes

const limiter = new RateLimit({
    store: limiterStore,
    windowMs: windowTime,
    max: 100, // 100 requests per 15 minutes
});

module.exports = limiter;
