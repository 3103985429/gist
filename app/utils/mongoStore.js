const MongoStore = require('rate-limit-mongo');
const winston = require('winston');

const dbURI = process.env.MONGODB_URI;

const limiterStore = new MongoStore({
    uri: dbURI,
    collectionName: 'expressRateRecords',
    errorHandler: winston.error,
});


const slowDownStore = new MongoStore({
    uri: dbURI,
    collectionName: 'slowDown',
    errorHandler: winston.error,
});


module.exports = {
    limiterStore,
    slowDownStore,
};
