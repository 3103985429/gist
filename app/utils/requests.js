const winston = require('winston');

const { PublicError } = require('../libs/system/errors');

module.exports = {
    handleError,
};

/**
 * Returns true when an error occurred.
 * @param {Error} err
 * @param {Object} res
 * @param {String=} message
 * @returns {Boolean}
 */
function handleError(err, res, message = 'An unexpected error happened during request.') {
    if (err) {
        const code = err.statusCode || 500;
        let responseMessage = message;

        if (err instanceof PublicError) {
            responseMessage = err.message;
        } else if (['development', 'testing'].includes(process.env.NODE_ENV)) {
            responseMessage = `Non-public error: ${err.message}`;
        }

        winston.error(message, err);
        res.status(code).json({ error: responseMessage });

        return true;
    }

    return false;
}
