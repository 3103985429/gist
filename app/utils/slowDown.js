const winston = require('winston');
const SlowDown = require('express-slow-down');
const { slowDownStore } = require('./mongoStore');

const windowTime = 15 * 60 * 1000; //15 minutes

const slowDown = (delayAfter = 20, windowTimeMs = windowTime) => new SlowDown({
    store: slowDownStore,
    windowMs: windowTimeMs,
    delayAfter,
    delayMs: 500,
    onLimitReached: (req) => {
        winston.error(`Throttled path: ${req.route.path}`);
        winston.error(`Throttled ip: ${req.headers['x-forwarded-for'] || req.connection.remoteAddress}`);
    }
});

module.exports = slowDown;
