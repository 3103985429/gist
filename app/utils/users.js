const Users = require('../models/users');

module.exports = {
    getUser,
};

/**
 * @param {String|ObjectId} id
 * @returns {Promise<Object>}
 */
async function getUser(id) {
    return await Users
        .findById(id)
        .lean()
        .exec();
}
