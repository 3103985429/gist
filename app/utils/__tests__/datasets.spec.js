import * as datasets from '../datasets';
import { hydrateDb } from '../../../internals/testing/backend/utils';
import { pokemon1 } from '../../../internals/testing/backend/fixtures/datasets';

describe('Datasets Utils', () => {
    beforeEach(async () => {
        await hydrateDb();
    });

    describe('when #getDataset is called', () => {
        it('should return a dataset', async () => {
            const result = await datasets.getDataset(pokemon1._id);

            expect(result).toMatchSnapshot();
        });
    });

    describe('when #hasValidDatasetSource is called', () => {
        it('should return true with a valid data source', () => {
            const datasourceDescription = {
                pipedrive: 'temp1'
            };

            expect(datasets.hasValidDatasetSource(datasourceDescription)).toBeTruthy();
            expect(datasets.hasValidDatasetSource(pokemon1)).toBeTruthy();
        });

        it('should return false with an invalid data source', () => {
            expect(datasets.hasValidDatasetSource({})).toBeFalsy();
        });
    });
});
