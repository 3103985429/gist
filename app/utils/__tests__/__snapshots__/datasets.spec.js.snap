// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datasets Utils when #getDataset is called should return a dataset 1`] = `
Object {
  "JSONPath": "undefined",
  "__v": 0,
  "_id": "5a42c26e303770207ce8f83b",
  "_otherSources": Array [],
  "_team": Object {
    "__v": 0,
    "_id": "5a42bf30d3bc0b20648290e1",
    "admin": Array [
      "5a42bf07d3bc0b20648290e0",
    ],
    "brandColor": Object {
      "label": "#2A2A2A",
      "select": "#2A2A2A",
    },
    "colorPalette": Array [
      "#1B6DFC",
      "#0CB7FA",
      "#FA79FC",
      "#FB3705",
      "#F9D307",
      "#FA9007",
      "#7D5807",
      "#B2E606",
      "#069908",
      "#7B07FD",
    ],
    "createdAt": 2017-12-26T21:29:20.384Z,
    "datasourceDescriptions": Array [
      "5a42c26e303770207ce8f83b",
      "5b033d64ef29594d18bb18aa",
    ],
    "font": "calibre",
    "hasEnterpriseLicense": false,
    "pages": Array [
      "5ac26b456f8073766713bcee",
      "5ac26b3d6f8073766713bced",
      "5ac26b506f8073766713bcef",
      "5ac26b636f8073766713bcf0",
      "aaa26b456f8073766713bcee",
    ],
    "sites": Array [],
    "subdomain": "glitter",
    "superTeam": true,
    "title": "Glitter",
    "updatedAt": 2017-12-26T21:43:10.717Z,
  },
  "author": Object {
    "__v": 0,
    "_articleEditors": Array [],
    "_articleViewers": Array [],
    "_editors": Array [],
    "_id": "5a42bf07d3bc0b20648290e0",
    "_siteEditors": Array [],
    "_siteViewers": Array [],
    "_team": Array [
      "5a42bf30d3bc0b20648290e1",
    ],
    "_viewers": Array [],
    "activated": true,
    "active": true,
    "canCreateNewArticle": Array [],
    "canCreateNewSite": Array [],
    "canCreateNewViz": Array [],
    "createdAt": 2017-12-26T21:28:39.946Z,
    "defaultLoginTeam": "5a42bf30d3bc0b20648290e1",
    "email": "user@schemadesign.com",
    "firstName": "Maitland",
    "hash": "dc449e96c930ca46653ec867c448719fab39dd85c4b91c1643d2faad140f24e711da13133e34ee4da0d3a5a2ca154edd4bfaeb9cf80cb9af57bce4d7ea1b94ce",
    "lastName": "McConnell",
    "lockUntil": 1,
    "loginAttempts": 0,
    "provider": "local",
    "salt": "test",
    "sampleImported": false,
    "signedUp": false,
    "updatedAt": 2017-12-26T21:36:31.420Z,
  },
  "brandColor": Object {
    "accent": "#005CB5",
  },
  "clones": 0,
  "columns": Array [
    Object {
      "data_type": "Text",
      "name": "Image URL",
      "operation": "ToString",
    },
    Object {
      "data_type": "Number",
      "name": "Max HP",
      "operation": "ToInteger",
    },
    Object {
      "data_type": "Number",
      "name": "Max CP",
      "operation": "ToInteger",
    },
    Object {
      "data_type": "Text",
      "name": "Type 2",
      "operation": "ToString",
    },
    Object {
      "data_type": "Text",
      "name": "Type 1",
      "operation": "ToString",
    },
    Object {
      "data_type": "Text",
      "name": "Name",
      "operation": "ToString",
    },
    Object {
      "data_type": "Number",
      "name": "Pokemon No_",
      "operation": "ToInteger",
    },
  ],
  "contentEdited": false,
  "createdAt": 2017-12-26T21:43:10.707Z,
  "createdFields": Array [],
  "customFieldsToProcess": Array [],
  "dirty": 0,
  "fe_displayTitleOverrides": Object {
    "Type 1": "Type1_overridden",
  },
  "fe_excludeFields": Object {
    "Image URL": false,
    "Max CP": false,
    "Max HP": false,
    "Name": false,
    "Pokemon No_": false,
    "Type 1": false,
    "Type 2": false,
  },
  "fe_excludeFieldsObjDetail": Object {
    "Image URL": false,
    "Max CP": false,
    "Max HP": false,
    "Name": false,
    "Pokemon No_": false,
    "Type 1": false,
    "Type 2": false,
  },
  "fe_fieldDisplayOrder": Array [],
  "fe_filters": Object {
    "default": Object {},
    "excludeFields": Array [],
    "fabricated": Array [],
    "fieldsAsRangeSlider": Array [],
    "fieldsCommaSeparatedAsIndividual": Array [],
    "fieldsMultiSelectable": Array [],
    "fieldsNotAvailable": Array [],
    "fieldsSortableByInteger": Array [],
    "fieldsSortableInReverseOrder": Array [],
    "keywords": Array [],
    "valuesToExcludeByOriginalKey": Object {
      "_all": Array [],
    },
  },
  "fe_image": Object {
    "detailView": true,
    "field": "",
    "gallery": true,
    "overwrite": false,
    "scraped": false,
    "timeline": true,
  },
  "fe_listed": true,
  "fe_nestedObject": Object {
    "criteria": Object {},
    "fields": Array [],
  },
  "fe_viewOptions": Object {
    "enableAccessibility": false,
    "enableLastUpdated": false,
    "filtersRail": true,
    "fullscreenExpand": true,
    "insightsExplorer": true,
    "insightsExplorerEmbed": true,
    "search": true,
    "viewControls": true,
    "viewInteractivity": true,
  },
  "fe_views": Object {
    "default_view": "gallery",
    "views": Object {
      "gallery": Object {
        "defaultSortByColumnName": "Image URL",
        "visible": true,
      },
      "pieChart": Object {
        "defaultGroupByColumnName": "Name",
        "visible": true,
      },
      "pieSet": Object {
        "defaultAggregateByColumnName": "Max HP",
        "defaultChartByColumnName": "Type 1",
        "defaultGroupByColumnName": "Type 2",
        "fieldsNotAvailableAsAggregateByColumns": Array [],
        "fieldsNotAvailableAsChartByColumns": Array [],
        "fieldsNotAvailableAsGroupByColumns": Array [],
        "valuesToExcludeByOriginalKey": Array [],
        "visible": true,
      },
      "timeline": Object {
        "visible": true,
      },
      "treemap": Object {
        "defaultAggregateByColumnName": "Max HP",
        "defaultChartByColumnName": "Name",
        "defaultGroupByColumnName": "Type 1",
        "fieldsNotAvailableAsAggregateByColumns": Array [],
        "fieldsNotAvailableAsChartByColumns": Array [],
        "fieldsNotAvailableAsGroupByColumns": Array [],
        "visible": true,
      },
    },
  },
  "fe_visible": true,
  "fileName": "pokemonGO.csv",
  "firstImport": 0,
  "format": "CSV",
  "importRevision": 1,
  "imported": true,
  "includeEmptyFields": true,
  "isPublic": true,
  "jobId": 0,
  "lastImportInitiatedBy": null,
  "lastImportTriggeredBy": "manual",
  "objectTitle": "Pokemon No_",
  "openDownload": false,
  "preserveEditedData": true,
  "previous_datasets": Array [],
  "raw_rowObjects_coercionScheme": Object {
    "Image URL": Object {
      "operation": "ToString",
    },
    "Max CP": Object {
      "operation": "ToInteger",
    },
    "Max HP": Object {
      "operation": "ToInteger",
    },
    "Name": Object {
      "operation": "ToString",
    },
    "Pokemon No_": Object {
      "operation": "ToInteger",
    },
    "Type 1": Object {
      "operation": "ToString",
    },
    "Type 2": Object {
      "operation": "ToString",
    },
  },
  "relationshipFields": Array [],
  "replaced": false,
  "replacement": false,
  "sample": false,
  "skipImageScraping": false,
  "state": "approved",
  "tabDestination": 0,
  "title": "Pokemon",
  "uid": "pokemon",
  "updatedAt": 2017-12-26T21:44:14.849Z,
  "updatedBy": "5a42bf07d3bc0b20648290e0",
  "updatedContent": Object {
    "created": Object {},
    "edited": Object {},
    "new": true,
    "publishedStatus": Object {},
  },
  "urls": Array [],
}
`;
